river/TextInput.zig

```diff
@@ -0,0 +1,368 @@
+    const node = util.gpa.create(std.TailQueue(TextInput).Node) catch return;
+    node.data = .{
+        .relay = relay,
+        .wlr_text_input = wlr_text_input,
+    };
+
+    node.data.wlr_text_input.events.enable.add(&node.data.enable);
+    node.data.wlr_text_input.events.commit.add(&node.data.commit);
+    node.data.wlr_text_input.events.disable.add(&node.data.disable);
+    node.data.wlr_text_input.events.destroy.add(&node.data.destroy);
+
+    relay.text_inputs.append(node);
+    log.debug("new text input on seat {s}", .{relay.seat.wlr_seat.name});
+}
+
+fn handleEnable(listener: *wl.Listener(*wlr.TextInputV3), wlr_text_input: *wlr.TextInputV3) void {
+    const self = @fieldParentPtr(TextInput, "enable", listener);
+
+    const input_method = self.relay.input_method orelse return;
+    input_method.sendActivate();
+
+    self.relay.sendInputMethodState(self.wlr_text_input);
+}
+
+fn handleCommit(listener: *wl.Listener(*wlr.TextInputV3), wlr_text_input: *wlr.TextInputV3) void {
+    const self = @fieldParentPtr(TextInput, "commit", listener);
+    if (self.relay.input_method == null) {
+        log.debug("text input committed but input method is gone", .{});
+        return;
+    }
+    self.relay.sendInputMethodState(self.wlr_text_input);
+}
+
+fn handleDisable(listener: *wl.Listener(*wlr.TextInputV3), wlr_text_input: *wlr.TextInputV3) void {
+    const self = @fieldParentPtr(TextInput, "disable", listener);
+    if (self.wlr_text_input.focused_surface == null) {
+        log.debug("disabling text_input, but no longer focused", .{});
+        return;
+    }
+    self.relay.disableTextInput(self);
+}
+
+fn handleDestroy(listener: *wl.Listener(*wlr.TextInputV3), wlr_text_input: *wlr.TextInputV3) void {
+    const self = @fieldParentPtr(TextInput, "destroy", listener);
+    const node = @fieldParentPtr(std.TailQueue(TextInput).Node, "data", self);
+
+    if (self.wlr_text_input.current_enabled) self.relay.disableTextInput(self);
+
+    node.data.setPendingFocusedSurface(null);
+
+    self.enable.link.remove();
+    self.commit.link.remove();
+    self.disable.link.remove();
+    self.destroy.link.remove();
+
+    self.relay.text_inputs.remove(node);
+    util.gpa.destroy(node);
+}
+
+fn handlePendingFocusedSurfaceDestroy(listener: *wl.Listener(*wlr.Surface), surface: *wlr.Surface) void {
+    const self = @fieldParentPtr(TextInput, "pending_focused_surface_destroy", listener);
+    // TODO: Currently often wrong so cause panic, need to fix it before using assert.
+    // assert(self.pending_focused_surface == surface);
+
+    self.pending_focused_surface = null;
+    self.pending_focused_surface_destroy.link.remove();
+}
+
+fn setPendingFocusedSurface(self: *TextInput, wlr_surface: ?*wlr.Surface) void {
+    self.pending_focused_surface = wlr_surface;
+    if (self.pending_focused_surface) |surface| {
+        surface.events.destroy.add(&self.pending_focused_surface_destroy);
+    }
```

Comment:
Need to remove the current destroy listener if there is a pending focused surface. This should fix the assertion failure you were getting in `handlePendingFocusedSurfaceDestroy()`.